digits/model/tasks/caffe_train.py

```diff
@@ -860,7 +860,10 @@ def _pycaffe_args(self, gpu_id):
             solver_type = solver_type_mapping[self.solver_type]
         except KeyError:
             raise ValueError("Unknown solver type {}.".format(self.solver_type))
-        gpu_script = "caffe.set_device({id});".format(id=gpu_id) if gpu_id else ""
+        if gpu_id:
+            gpu_script = "caffe.set_device({id});caffe.set_mode_gpu();".format(id=gpu_id)
+        else:
+            gpu_script = "caffe.set_mode_gpu();"
```

Comment:
> Those two conditions do not set mode to CPU in regular path of task_arguments. I tried to match that behavior in _pycaffe_args.

Technically this is how we do it for Caffe training:
https://github.com/NVIDIA/DIGITS/blob/v4.0.0-rc.1/digits/model/tasks/caffe_train.py#L422-L427

But that's pretty sloppy. In this case, I think it makes sense to set the mode to CPU if there is no GPU assigned.